src/Perl6/Optimizer.nqp

```diff
@@ -260,28 +260,47 @@ my class Symbols {
         return 0;
     }
 
+    # method find_in_setting($symbol) {
+    #     if !nqp::defined($!SETTING) {
+    #         my int $i := +@!block_stack;
+    #         while $i > 0 && !nqp::defined($!SETTING) {
+    #             $i := $i - 1;
+    #             my $block := @!block_stack[$i];
+    #             my %sym := $block.symbol("!CORE_MARKER");
+    #             if +%sym {
+    #                 $!SETTING := $block;
```

Comment:
NQP does support `last`.  Put it here, then you could remove the `!defined($!SETTING)` from the while condition for more simple code.